go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2120,6 +2123,261 @@ func (s *VtctldServer) UpdateCellsAlias(ctx context.Context, req *vtctldatapb.Up
+	}
+
+	sort.Strings(shards)
+	if len(shards) == 1 {
+		validateSchemaShardReq := &vtctldatapb.ValidateSchemaShardRequest{
+			Keyspace:       req.Keyspace,
+			ExcludeTables:  req.ExcludeTables,
+			Shard:          shards[0],
+			IncludeVSchema: req.IncludeVSchema,
+			IncludeViews:   req.IncludeViews,
+		}
+
+		if _, err := s.ValidateSchemaShard(ctx, validateSchemaShardReq); err != nil {
+			return nil, vterrors.Wrapf(err, "ValidateSchemaShard(%+v) failed: %v", validateSchemaShardReq, err)
+		}
+		return &vtctldatapb.ValidateSchemaKeyspaceResponse{}, nil
+	}
+
+	var (
+		referenceSchema *tabletmanagerdatapb.SchemaDefinition
+		referenceAlias  *topodatapb.TabletAlias
+		rec             concurrency.AllErrorRecorder
+		wg              sync.WaitGroup
+	)
+
+	if req.IncludeVSchema {
+		validateVSchemaReq := &vtctldatapb.ValidateVSchemaRequest{
+			Keyspace:      req.Keyspace,
+			Shards:        shards,
+			IncludeViews:  req.IncludeViews,
+			ExcludeTables: req.ExcludeTables,
+		}
+
+		if _, err := s.ValidateVSchema(ctx, validateVSchemaReq); err != nil {
+			return nil, vterrors.Wrapf(err, "ValidateVSchema(%v) failed: %v", validateVSchemaReq, err)
+		}
+	}
+
+	// Diff tablets against the reference shard schema
+	for _, shard := range shards {
+		si, err := s.ts.GetShard(ctx, req.Keyspace, shard)
+		if err != nil {
+			rec.RecordError(fmt.Errorf("GetShard(%v, %v) failed: %v", req.Keyspace, shard, err))
+			continue
+		}
+
+		if !si.HasPrimary() {
+			if !req.SkipNoPrimary {
+				rec.RecordError(fmt.Errorf("no primary in shard %v/%v", req.Keyspace, shard))
+			}
+			continue
+		}
+
+		if referenceSchema == nil {
+			referenceAlias = si.Shard.PrimaryAlias
+			log.Infof("Gathering schema for reference primary %v", topoproto.TabletAliasString(referenceAlias))
+
+			getTabletReq := &vtctldatapb.GetTabletRequest{
+				TabletAlias: referenceAlias,
+			}
+
+			getTabletResp, err := s.GetTablet(ctx, getTabletReq)
+			if err != nil {
+				return nil, vterrors.Wrapf(err, "GetTablet(%v) failed: %v", topoproto.TabletAliasString(referenceAlias), err)
+			}
+
+			referenceSchema, err = s.tmc.GetSchema(ctx, getTabletResp.GetTablet(), nil, req.ExcludeTables, req.IncludeViews)
+			if err != nil {
+				return nil, vterrors.Wrapf(err, "GetSchema(%v, nil, %v, %v) failed: %v", referenceAlias, req.ExcludeTables, req.IncludeViews, err)
+			}
+		}
+
+		aliases, err := s.ts.FindAllTabletAliasesInShard(ctx, req.Keyspace, shard)
+		if err != nil {
+			rec.RecordError(fmt.Errorf("FindAllTabletAliasesInShard(%v, %v) failed: %v", req.Keyspace, shard, err))
+			continue
+		}
+
+		for _, alias := range aliases {
+			// Don't diff schemas for self
+			if referenceAlias == alias {
+				continue
+			}
+			wg.Add(1)
+			go func(alias *topodatapb.TabletAlias) {
+				defer wg.Done()
+				if err := s.diffSchema(ctx, referenceSchema, referenceAlias, alias, req.ExcludeTables, req.IncludeViews); err != nil {
+					rec.RecordError(err)
+				}
+			}(alias)
+		}
+	}
+	wg.Wait()
```

Comment:
whitespace is off here. refer to the style of the other functions in this file please